### ISSUE TITLE
Questions about federated learning in Inverting Gradients

### ISSUE THREAD

**Issue by Rahn80643** [JonasGeiping/breaching#5]
Dear Geiping,

Thank you for providing the implementation of paper ["Inverting Gradients - How easy is it to brak privacy in federated learning?"](https://arxiv.org/pdf/2003.14053.pdf). After reading the paper and code, I have some questions regarding to the implementation:

1. The paper mentions if a neural network contains a fully-connected layer, it is possible to reconstruct the input of the network from the network's gradients. I'm confuse about the description, does this means for any neural network containing fully-connected layer in the last, the input can be calculated by the overall gradients of the entire model, instead of only using the gradient of the last fully-connected layer?

2. In the experiment of trained VS. untrained networks in page 6 of the paper, may I ask that the untrained model is a pretrained model rather than the model with the weights randomly initialized? For example, in the experiment in Table 1, the untrained models means the models' weight values are initialized randomly, while the trained models were trained with CIFAR-10 dataset?

3. I tried to apply the federated learning part in inverting gradient, I do the following modification:
    1) change user to multiuser_aggregate in breaching/config/case/4_fedavg_small_scale.yaml
    2) set default_clients to 10 in breaching/case/data/CIFAR10.yaml

    However, I got the following error message after generating the dataset for training:
    "This user would have no data under the chosen partition, user id and number of clients.", which maybe caused the dataset is split more than 10 clients. Apart from the settings above, which files should I modify to apply multiple clients in federated learning combined with inverting gradients?

![image](https://user-images.githubusercontent.com/8951991/222628922-c24c4f44-765c-4254-ac25-48aacc2e127b.png)

Best Regards,
Rahn

**Comment by JonasGeiping** [comment]
Hi, 

1. The (average) input of any fullly-connected layer can be reconstructed. For a model that ends with such a layer, you can then recover the inputs to the last layer from the gradients of the last layer.
2. Untrained refers to a random initialization, not to a pretrained model.
3. If you want to have 1000 users aggregated, then you need to configure the dataset to be partitioned into 1000 users. See the dataset configuration at https://github.com/JonasGeiping/breaching/blob/a51fba6b637207adbfbbde3a117eed2391c515ce/breaching/config/case/data/CIFAR10.yaml#L31.
What about first testing the multi-user aggregate with 10 users, does it work for you in this case?

By the way, all of these config changes can be done at the cmd-line, no need to modify the configs yamls.

**Comment by Rahn80643** [comment]
Hi,
Thank you so much for your quick reply and explanation. After setting `cfg.case.user.user_range` in  multi-user aggregate, the training procedure could be completed without error.

In addition, I have two questions:
1. In class `UserMultiStep`, the amount of data for training is the same as `num_data_points`. Does this mean the client uses only a few data to train a model instead of all training data like CIFAR10? If the client uses all local data to train the model, the server can try to reconstruct all the images?
2. Function `reconstruct()` and `_cosine_sim()` aim to train gradients similar to the target gradients. I'm confuse about why as the number of images increase, the length or the size of the gradients still remain the same, and the images can be reconstructed in the same reconstruction process instead of reconstruct one by one? 

Best Regards

**Comment by JonasGeiping** [comment]
1. This really depends on what setting you want to model. How much data should each client (or with MultiUser, each  group of clients/users) use for the update?
2. The parameter gradient has the same dimension as the parameters, its size is independent of the amount of data used in the update. 

There also might be a small confusion here between what attack is possible, and what attack will succeed. The more data is aggregated into a user update, the harder it is for the attack to recover data in the honest-but-curious threat model that you are considering here. 
The optimization problem of the attacker can be formulated for any model and amount of user data, but solutions are not always original data points, this depends on model size and number of data points.

**Comment by Rahn80643** [comment]
Thank you for explanation, it seems the more clients' models in aggregation and the more data the client use, it would be harder for the attacker to reconstruct the data?

Best Regards

**Comment by JonasGeiping** [comment]
At least for this type of attack. With malicious-server attacks, using more data or more clients can also be defeated.

**Comment by Rahn80643** [comment]
Thank you for your explanation, it's helps me a lot understand this attack.

Best Regards
